src/masternodes/mn_checks.cpp

```diff
@@ -2037,6 +2044,101 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
         return Res::Ok();
     }
 
+    
+    Res operator()(const CLockDUSDMessage& obj) const {
+        auto res = CheckCustomTx();
+        if (!res)
+            return res;
+
+        if (!HasAuth(obj.owner)) {
+            return Res::Err("Transaction must have at least one input from owner");
+        }
+
+        if (static_cast<int>(height) < consensus.GrandCentralHeight) {
+            return Res::Err("LockDUSD only available after Grand Central");
+        }
+
+        const auto attributes = mnview.GetAttributes();
+        if (!attributes) {
+            return Res::Err("Attributes unavailable");
+        }
+
+        const auto paramID = ParamIDs::DFIP2211D;
+
+        //TODO: refactor to have lockperiods in gov "dfip2211d/locklimit/<months>", if key is there, locking is active with given limit
+        CDataStructureV0 activeKey{AttributeTypes::Param, paramID, DFIPKeys::Active};
+        CDataStructureV0 limit12Key{AttributeTypes::Param, paramID, DFIPKeys::LOCK_12_Limit};
+        CDataStructureV0 limit24Key{AttributeTypes::Param, paramID, DFIPKeys::LOCK_24_Limit};
+        if (!attributes->GetValue(activeKey, false) ||
+            !attributes->CheckKey(limit12Key) ||
+            !attributes->CheckKey(limit24Key)) {
+            return Res::Err("DFIP2211D not currently active");
+        }
+
+        if (obj.dusdIn <= 0) {
+            return Res::Err("Source amount must be more than zero");
+        }
+
+        //check locktime. must be 12 or 24
+        if(obj.lockTime != 12 || obj.lockTime != 24) {
+            return Res::Err("Can only lock for 12 or 24 months");
+        }
+        auto lockToken = mnview.GetToken(strprintf("DUSDL%i",obj.lockTime));
+        if(!lockToken) {
+            return Res::Err("Could not find lock token");
+        }
+        
+        auto dusdToken = mnview.GetToken("DUSD");
+        if(!dusdToken) {
+            return Res::Err("Could not find dusd token");
+        }
+
+        const auto contractAddressValue = GetFutureSwapContractAddress(SMART_CONTRACT_DFIP2211D);
+        if (!contractAddressValue) {
+            return contractAddressValue;
+        }
+
+        auto limit = attributes->GetValue(obj.lockTime == 12 ? limit12Key : limit24Key, CAmount{0});
+        auto resultedMint= SafeAdd(lockToken->second->minted, obj.dusdIn);
+        if (!resultedMint)
+            return (std::move(resultedMint));
+
+        if(*resultedMint.val > limit) {
+            return Res::Err("Limit reached for this lock token");
+        }
+
+        //do we need CalculateOwnerRewards here?
+        res = mnview.SubBalance(obj.owner, CTokenAmount{dusdToken->first, obj.dusdIn});
+        if (!res)
+            return res;
+        
+        auto half= DivideAmounts(obj.dusdIn,CAmount{2});
+        //half DUSD are burned, that amount DUSDLOCK is minted
+        //remaining half + DUSDLOCK added to pool
+        //sub minted or burn?
+        res = mnview.SubMintedTokens(dusdToken->first, half);
```

Comment:
SubMintedToknes are used to maintain the evidence of minted amount, but it doesn't burn anything. So you must also send those DUSD to burn address.

Reply:
do I? I substract it from the balance and reduce the minted. Thats basically the same thing that happens when they would be used to payback a loan. (its as if they have never been there at all)
Since they are replaced by locked_dusd I thought its better two "swap" minted for minted: reduce minted DUSD and increase minted lock.
if we send them to burn, we increase total amount (not net supply but total minted) in the system which feels "wrong" here.